messaging/src/main/java/org/axonframework/messaging/LambdaBasedMessageTypeResolver.java

```diff
@@ -0,0 +1,168 @@
+ * </ul>
+ *
+ * @author Mateusz Nowak
+ * @since 5.0.0
+ */
+public class LambdaBasedMessageTypeResolver implements MessageTypeResolver {
+
+    private final Map<Class<?>, MessageTypeResolver> resolvers;
+    private final MessageTypeResolver defaultResolver;
+
+    /**
+     * Private constructor used by the builder to create the resolver with the configured resolvers and default
+     * behavior.
+     *
+     * @param resolvers       The map of payload types to their specific resolvers
+     * @param defaultResolver The default resolver to use when no specific resolver is found, or null to throw an
+     *                        exception
+     */
+    private LambdaBasedMessageTypeResolver(Map<Class<?>, MessageTypeResolver> resolvers,
+                                           MessageTypeResolver defaultResolver) {
+        this.resolvers = resolvers;
+        this.defaultResolver = defaultResolver;
+    }
+
+    /**
+     * Starts building a new resolver and registers the first payload type with a custom resolver.
+     *
+     * @param payloadType The class to register the resolver for
+     * @param resolver    The resolver that resolves the payload type class to a {@link MessageType}
+     * @return The type resolver phase for further configuration
+     */
+    public static TypeResolverPhase on(Class<?> payloadType, MessageTypeResolver resolver) {
+        Map<Class<?>, MessageTypeResolver> initialResolvers = new HashMap<>();
+        initialResolvers.put(payloadType, resolver);
+        return new InternalTypeResolverPhase(initialResolvers, null);
+    }
+
+    /**
+     * Starts building a new resolver and registers the first payload type with a fixed message type.
+     *
+     * @param payloadType The class to register the fixed message type for
+     * @param messageType The message type to use for the given payload type
+     * @return The type resolver phase for further configuration
+     */
+    public static TypeResolverPhase on(Class<?> payloadType, MessageType messageType) {
+        return on(payloadType, payloadType1 -> messageType);
+    }
+
+    @Override
+    public MessageType resolve(Class<?> payloadType) {
+        var resolver = resolvers.get(payloadType);
+        if (resolver != null) {
+            return resolver.resolve(payloadType);
+        }
+        if (defaultResolver == null) {
+            throw new IllegalArgumentException("No resolver found for payload type [" + payloadType.getName() + "]");
+        }
+        return defaultResolver.resolve(payloadType);
+    }
+
+    /**
+     * Interface representing the phase where type resolvers are registered.
+     */
+    public interface TypeResolverPhase {
+
+        /**
+         * Registers a resolver that returns a {@link MessageType} for the given payload type.
+         *
+         * @param payloadType The class to register the resolver for
```

Comment:
Missing more dots on the `@param`, `@return`, and `@throws` statements further down.